flink-tests/src/test/java/org/apache/flink/test/recovery/BatchFineGrainedRecoveryITCase.java

```diff
@@ -199,6 +371,76 @@ private boolean toss() {
 		}
 	}
 
+	private enum StaticMapFailureTracker {
+		;
+
+		private static final List<AtomicInteger> mapRestarts = new ArrayList<>(MAP_NUMBER);
+		private static final List<AtomicInteger> expectedMapRestarts = new ArrayList<>(MAP_NUMBER);
+		private static final List<AtomicInteger> mapFailures = new ArrayList<>(MAP_NUMBER);
+		private static final List<AtomicInteger> mapFailuresWithBacktracking = new ArrayList<>(MAP_NUMBER);
+		private static final AtomicReference<Throwable> unrelatedFailure = new AtomicReference<>(null);
+
+		private static void reset() {
+			mapRestarts.clear();
+			expectedMapRestarts.clear();
+			mapFailures.clear();
+			mapFailuresWithBacktracking.clear();
+		}
+
+		private static int addNewMap() {
+			mapRestarts.add(new AtomicInteger(0));
+			expectedMapRestarts.add(new AtomicInteger(1));
+			mapFailures.add(new AtomicInteger(0));
+			mapFailuresWithBacktracking.add(new AtomicInteger(0));
+			return mapRestarts.size() - 1;
+		}
+
+		private static void mapRestart(int index) {
+			mapRestarts.get(index).incrementAndGet();
+		}
+
+		private static void mapFailure(int index) {
+			mapFailures.get(index).incrementAndGet();
+			expectedMapRestarts.get(index).incrementAndGet();
+		}
+
+		private static void mapFailureWithBacktracking(int index) {
+			mapFailuresWithBacktracking.get(index).incrementAndGet();
+			IntStream.range(0, index + 1).forEach(i -> expectedMapRestarts.get(i).incrementAndGet());
+			IntStream.range(0, index + 1).forEach(i -> expectedMapRestarts.get(i).incrementAndGet());
+		}
+
+		private static void unrelatedFailure(Throwable failure) {
+			unrelatedFailure.getAndUpdate(t -> {
+				// wrap to avoid side effects in the atomic updater (multiple failure.addSuppressed calls)
```

Comment:
instead of potentially wrapping this multiple times a simple `synchronized` block may be a better option